src/NHibernate/Async/Event/Default/DefaultLockEventListener.cs

```diff
@@ -0,0 +1,93 @@
+﻿//------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by AsyncGenerator.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+
+using System;
+using NHibernate.Engine;
+using NHibernate.Persister.Entity;
+using NHibernate.Proxy;
+
+namespace NHibernate.Event.Default
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	/// <content>
+	/// Contains generated async methods
+	/// </content>
+	public partial class DefaultLockEventListener : AbstractLockUpgradeEventListener, ILockEventListener
+	{
+		/// <summary>Handle the given lock event. </summary>
+		/// <param name="event">The lock event to be handled.</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work</param>
+		public virtual Task OnLockAsync(LockEvent @event, CancellationToken cancellationToken)
+		{
+			if (@event.Entity == null)
+			{
+				throw new NullReferenceException("attempted to lock null");
+			}
+
+			if (@event.LockMode == LockMode.Write)
+			{
+				throw new HibernateException("Invalid lock mode for lock()");
+			}
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			async Task InternalOnLockAsync()
```

Comment:
>Is there a consensus about local function usage?

Please take a look [here](https://github.com/maca88/AsyncGenerator/issues/1#issuecomment-295933941). The generator will create a tail local function or method ([configurable](https://github.com/maca88/AsyncGenerator/blob/master/Source/AsyncGenerator.Tests/ExternalProjects/NHibernate/Fixture.cs#L98)) when there is at least one precondtion in the beginning of the method.
>  I rather put them after the holding function last return statement

I will put the local function declaration after the return statement.